packages/graph/hash_graph/lib/graph/src/shared/subgraph/mod.rs

```diff
@@ -29,4 +38,56 @@ impl Subgraph {
             depths,
         }
     }
+
+    fn entry<R: Record>(
+        &mut self,
+        edition_id: &impl SubgraphIndex<R>,
+    ) -> RawEntryMut<R::EditionId, R, RandomState> {
+        edition_id.subgraph_entry(self)
+    }
+
+    pub fn insert<R: Record>(&mut self, record: R) -> Option<R> {
+        match self.entry(record.edition_id()) {
+            RawEntryMut::Occupied(mut entry) => Some(entry.insert(record)),
+            RawEntryMut::Vacant(entry) => {
+                entry.insert(record.edition_id().clone(), record);
+                None
+            }
+        }
+    }
+
+    /// Looks up a single [`Record`] in the subgraph or reads it from the [`Store`] and inserts it
+    /// if it is not yet in the subgraph.
+    ///
+    /// # Errors
+    ///
+    /// - Returns an error if the [`Record`] could not be read from the [`Store`].
+    ///
+    /// [`Store`]: crate::store::Store
+    pub async fn get_or_read<'r, R: Record + Sync + 'r>(
+        &'r mut self,
+        store: &impl Read<R>,
+        edition_id: &R::EditionId,
+    ) -> Result<&'r R, QueryError> {
+        Ok(match self.entry(edition_id) {
+            RawEntryMut::Occupied(entry) => entry.into_mut(),
+            RawEntryMut::Vacant(entry) => {
+                entry
+                    .insert(
+                        edition_id.clone(),
+                        store
+                            .read_one(&R::create_filter_for_edition_id(edition_id))
+                            .await?,
+                    )
+                    .1
+            }
+        })
+    }
+}
+
+pub trait SubgraphIndex<R: Record>: Clone + Eq + Hash + Into<GraphElementEditionId> {
```

Comment:
Can we have some documentation for this please

Reply:
Added a docstring to the trait and the function in 153f95a3bec949cda0ba9971c6179c4e8e08300c. I think the most important thing in the documentation is to state, that the operation will be performed on the vertices depending on `R`.